### PR TITLE
Adds funding for lead schools and refactors all funding

### PR DESCRIPTION
**Pull request by ralph-hawkins**
Lead school monthly payments:

localhost_3000_funding_monthly-payments(screenshot) (2)

Lead school salaries and grants:

localhost_3000_funding_annual-payments-lead-school(screenshot)

Accredited provider monthly payments:

localhost_3000_funding_monthly-payments(screenshot) (3)

Accredited provider bursaries and scholarships:

localhost_3000_funding_annual-payments-scitt(screenshot)

### CODE REVIEW

File: app/filters/app-misc.js [hunk 1]
```diff
@@ -162,42 +163,64 @@ filters.includes = (route, string) =>{
...
+    ['numberOfTtraineesNoBursaryAwarded', "amountUGIttOrTier4EyItt"],
+    ['numberOfTtraineesScholarship', "amountScholarship"]
+  ]
+
+  let firstRow = array[0]
+
+  pairs.forEach(pair => {
+    let attributeIsMissing = ! (firstRow[pair[0]]) 
+    if (attributeIsMissing) {
+      console.log(`${pair[0]} is not present!`)
```
Comment (edwardhorsford) [inline app/filters/app-misc.js]: Delete

**Comment by dankmitchell** [comment]
[Funding — lead schools](https://trello.com/c/xXoKTXGH/3173-funding-lead-schools)

File: app/filters/app-misc.js [hunk 2]
```diff
@@ -162,42 +163,69 @@ filters.includes = (route, string) =>{
...
+    typesOfFunding.push("scholarships")
   }
+  return typesOfFunding
+}
 
-  return fundingStreams
+filters.formatYearRange = (string) => {
+  return string
+    .replace(/(\d{4})\/(\d{2})/, '$1&nbsp;to&nbsp;$2')
+    .replace(/(\d{2})\/(\d{2})/, '20$1&nbsp;to&nbsp;$2');
```
Comment (edwardhorsford) [inline app/filters/app-misc.js:231]: Could you add a comments for this? and an example?

File: app/filters/objects.js [hunk 3]
```diff
@@ -158,6 +165,26 @@ filters.deleteBlankAttributes = (dictionary) => {
...
+    let match = array.every(item => {
+      return item[attribute] == params.value
+    })
+    if (!match){
+      filteredAttributes.push(attribute)
+    }
+  })
+  return filters.keepAttributes(array, filteredAttributes)
+}
+
```
Comment (edwardhorsford) [inline app/filters/objects.js]: Could you add a comment / description for this? maybe an example?

File: app/filters/objects.js [hunk 4]
```diff
@@ -175,6 +202,30 @@ filters.removeWhere = (arr, key, compare) => {
   })
   return filtered
 }
+
+// Filter results for only those containing attribute and value
```
Comment (edwardhorsford) [inline app/filters/objects.js]: I think this comment needs updating - we copied it from elsewhere

File: app/filters/objects.js [hunk 5]
```diff
@@ -175,6 +202,30 @@ filters.removeWhere = (arr, key, compare) => {
...
+  let filtered = arr.filter(item => {
+    return compare.some(string => {
+      return _.get(item, key) && _.get(item, key).includes(string)
+    })
+    return compare.includes(_.get(item, key))
+  })
+  return filtered
+}
+
+filters.whereDoesNotInclude = (arr, key, compare) => {
```
Comment (edwardhorsford) [inline app/filters/objects.js:278]: Add comment

File: app/filters/strings.js [hunk 6]
```diff
@@ -68,7 +68,9 @@ filters.currency = input => {
   function numberWithCommas(x) {
     return x.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",");
   }
-  if (input) return `£${numberWithCommas(input)}`
+  if      ( input > 0 ) { return `£${numberWithCommas(input)}` }
+  // makes negative number positive and puts minus sign in front of £
+  else if ( input < 0 ) { return `–£${numberWithCommas(input * -1 )}` }
```
Comment (edwardhorsford) [inline app/filters/strings.js]: I think we might want to either check that `input` is numeric or else convert to numeric

File: app/filters/strings.js [hunk 7]
```diff
@@ -87,6 +89,24 @@ filters.markDown = input => {
...
+    return string.startsWith(target)
+  } else {
+    return false
+  }
+}
+
+// Replaces "&" with "and"
+filters.ampersandToAnd = (string) => {
+  return string.replace(/&/g,"and")
+}
```
Comment (edwardhorsford) [inline app/filters/strings.js]: I wonder if we could just use the built-in replace filter for this?

File: app/views/funding/annual-payments-lead-school.html [hunk 8]
```diff
@@ -0,0 +1,85 @@
+{% extends "layout.html" %}
+
+{% block content %}
+
+  {% set pageHeading = "Funding 2021 to 2022" %}
+  {% set tabName = "Salares and grants" %}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-lead-school.html]: *Salares* lol

File: app/views/funding/annual-payments-lead-school.html [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+  {% set activeTabMonthly = false %}
+  {% set activeTabYearlyScitt = false %}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-lead-school.html]: I don't think you need to set these?

File: app/views/funding/annual-payments-lead-school.html [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    {% endif %}
+  {% endfor %}
+  {% set row = [
+    { },
+    { },
+    { },
+    { },
+    { text: total | currency,
+      format: "numeric" }
+  ] %}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-lead-school.html]: Unclear what this is for - as it looks a bit weird, can you add a comment?

File: app/views/funding/annual-payments-scitt.html [hunk 11]
```diff
@@ -0,0 +1,298 @@
+{% extends "layout.html" %}
+
+{% block content %}
+
+  {% set pageHeading = "Funding 2021 to 2022" %}
+
+  {% set yearPaymentsTabName = "" | typesOfFunding | joinify  | sentenceCase %}
+  {% set activeTabMonthly = false %}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-scitt.html]: Delete

File: app/views/funding/annual-payments-scitt.html [hunk 12]
```diff
@@ -0,0 +1,298 @@
...
+  {% set activeTabYearlyScitt = true %}
+  {% set activeTabYearlyLeadSchool = false %}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-scitt.html]: Delete

File: app/views/funding/annual-payments-scitt.html [hunk 13]
```diff
@@ -0,0 +1,298 @@
...
+        <a href="#" class="govuk-link--no-visited-state">Export {{ yearPaymentsTabName | lower }}</a>
+      </p>
+    </div>
+  </div>
+
+  {% set dataSource = data.funding.annualFundingScitts %}
+  {% set dataSource = dataSource | sort(attribute = "subject") %}
+
+  {# ITT Bursaries #}
+  {# to do – sort alphabetically by subject #}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-scitt.html]: Isn't this sorted by subject in line 28?

File: app/views/funding/annual-payments-scitt.html [hunk 14]
```diff
@@ -0,0 +1,298 @@
...
+              classes: "app-table__column-20"
+            }
+          ],
+          rows: eyIttBursaryRows
+        }) }}
+
+      </div>
+    </div>
+
+    {# Not showing this for now as its inferred #}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-scitt.html]:
```suggestion
    {# Not showing this for now as it’s inferred #}
```

File: app/views/funding/monthly-payments.html [hunk 15]
```diff
@@ -1,150 +1,104 @@
...
-    ]
-  }) }}
+  {% if accessLevel == 'accreditingProvider' %}
+    {% set dataSource = data.funding.monthlyFundingScitts %}
+  {% else %}
+    {% set dataSource = data.funding.monthlyFundingLeadSchools %}
+  {% endif %}
+  {% set dataSource = dataSource | sort(attribute = "description") %}
 
+  {# TODO #}
```
Comment (edwardhorsford) [inline app/views/funding/monthly-payments.html:19]: What is TODO?

File: app/views/funding/monthly-payments.html [hunk 16]
```diff
@@ -1,150 +1,104 @@
...
+      {% set rowsFuture = rowsFuture | push(row) %}
+    {% endif %}
   {% endfor %}
 
   {% set summaryRows = [] %}
-  {% if futurePaymentsTotal != 0 %}
-    {% set summaryRowItems = [{ text: "Payments to " + currentMonth("word") + " " + currentYear() }, { text: pastPaymentsTotal | currency, format: "numeric" }] %}
+  {% if thisMonth != 11 %}
+    {% set summaryRowItems = [
+      { text: "Payments to " + currentMonth("word") + " " + currentYear() },
```
Comment (edwardhorsford) [inline app/views/funding/monthly-payments.html]: What is "word"?

**Review by edwardhorsford — CHANGES_REQUESTED** [review]

File: app/filters/strings.js (hunk 7 above)
Comment (ralph-hawkins) [reply app/filters/strings.js]: yeah, have deleted

File: app/views/funding/annual-payments-lead-school.html (hunk 10 above)
Comment (ralph-hawkins) [reply app/views/funding/annual-payments-lead-school.html]: Realised this should be done with colSpan.

File: app/views/funding/monthly-payments.html (hunk 16 above)
Comment (ralph-hawkins) [reply app/views/funding/monthly-payments.html]: the name of the month — I've relabelled as "nameOfMonth"

File: app/filters/strings.js (hunk 6 above)
Comment (ralph-hawkins) [reply app/filters/strings.js]: done — though i think its ok without?

File: app/filters/objects.js [hunk 17]
```diff
@@ -158,6 +165,27 @@ filters.deleteBlankAttributes = (dictionary) => {
   return newDictionary;
 }
 
+// Delete all params where all items match params
```
Comment (edwardhorsford) [inline app/filters/objects.js]: This comment doesn't really make any sense - could you give an example?

File: app/functions.js [hunk 18]
```diff
@@ -31,9 +31,9 @@ module.exports = function (env) {
     return moment().format('YYYY-MM-DD')
   }
 
-  // Get current month as "word" (eg December = 'December'), or number (eg December = 12)
+  // Get current month as "nameOfMonth" (eg December = 'December'), or number (eg December = 12)
   functions.currentMonth = (type) => {
```
Comment (edwardhorsford) [inline app/functions.js:41]: Should this have a default, making the param optional?

```suggestion
  functions.currentMonth = (type="number") => {
```

File: app/functions.js [hunk 19]
```diff
@@ -31,9 +31,9 @@ module.exports = function (env) {
     return moment().format('YYYY-MM-DD')
   }
 
-  // Get current month as "word" (eg December = 'December'), or number (eg December = 12)
+  // Get current month as "nameOfMonth" (eg December = 'December'), or number (eg December = 12)
```
Comment (edwardhorsford) [inline app/functions.js]:
```suggestion
  // Get current month as "nameOfMonth" (eg December = 'December'), or "number" (eg December = 12)
```

File: app/views/funding/annual-payments-lead-school.html [hunk 20]
```diff
@@ -19,30 +20,31 @@ <h2 class="govuk-heading-l">
...
+    { text: "Total",
+      classes: "app-table__column-20" }
   ] %}
 
   {% set total = 0 %}
   {% set salaryBodyRows = [] %}
   {% for item in dataSource %}
     {% if (item.numberOfTrainees > 0) and (item.amountPerTrainee > 0) %}
       {% set subjectTotal = (item.numberOfTrainees * item.amountPerTrainee) %}
+      {% set subjectAndRoute = item.subject + '<br><span class="app-nowrap">' + item.route | lower | sentenceCase + '</span>' %}
```
Comment (edwardhorsford) [inline app/views/funding/annual-payments-lead-school.html]: This would probably be clearer as a set block.

**Review by edwardhorsford — APPROVED** [review]

**Pull request merged**